### PR TITLE
Added parser, formatter, getters for date components

### LINKED ISSUE #3: create a `format` method in our BS class

### LINKED ISSUE #4: create a static `parse` method in our BS class

### LINKED ISSUE #8: Create `getYear` `getMonth` & `getDay` methods

### PR DESCRIPTION
**Pull request by blackkspydo**
### Description

- added parser to parse date components from date-string
- added formatter to format date components to a date-string
- added `.format()`, `.getYear()`, `.getMonth()`, `.getDay()` , `.toString()`  methods to `BikramSambat()` class

### Issue Reference
-  #8 
-  #4 
-  #3 

### Proposed Changes

- added parser to parse date components from date-string along with relevant test cases
- added formatter to format date components to a date-string along with relevant test cases
- added `.format()`, `.getYear()`, `.getMonth()`, `.getDay()` methods to `BikramSambat()` class

### Checklist

- [x] Lint rules are passed correctly
- [x] Code builds correctly
- [x] Created or updated tests for the new code
- [x] All existing and new tests passed
- [x] Pull request title is concise and descriptive.
- [x] Pull request is linked to the related issue.

### CODE REVIEW

File: .prettierrc.json [hunk 1]
```diff
@@ -14,4 +14,4 @@
   "tabWidth": 2,
   "trailingComma": "none",
   "useTabs": false
-}
+}
```
Comment (ashiishme) [inline .prettierrc.json]: I think this change is not necessary and can be removed to reduce the PR size.

File: src/format.test.ts [hunk 2]
```diff
@@ -0,0 +1,187 @@
+import { generateDateFormatOrder } from 'utils'
+import { format } from './format'
+import BikramSambat from './BikramSambat'
+describe('Date Formatting', () => {
```
Comment (ashiishme) [inline src/format.test.ts:4]: can you also cover some fail cases that causes error?

File: src/format.test.ts [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+    expect(result).toEqual('2079/08 15')
+  })
+  it('should format date with mismatched separators (MMMM DD YYYY)', () => {
+    const date = new BikramSambat('2079-8-15')
+    const result = format(date, 'MMMM DD YYYY')
+    expect(result).toEqual('Mangsir 15 2079')
+  })
+})
+
+describe('generate order array', () => {
```
Comment (ashiishme) [inline src/format.test.ts]: These test are not related to the format file.

File: src/utils/index.ts [hunk 4]
```diff
@@ -0,0 +1 @@
+export * from './generateDateFormatOrder'
```
Comment (ashiishme) [inline src/utils/index.ts]: directly export without index.ts file.

File: src/BikramSambat.ts [hunk 5]
```diff
@@ -1,25 +1,83 @@
...
+      this.day === undefined
+    ) {
+      return 'Invalid Date'
+    }
+    if (!isDayValid(this.year, this.month, this.day)) {
+      return 'Invalid Date'
+    }
+    // the month and day are padded with 0 if they are less than 10
+    const month = this.month <= 9 ? '0' + this.month : this.month
+    const day = this.day <= 9 ? '0' + this.day : this.day
```
Comment (aabishkaryal) [inline src/BikramSambat.ts]: what about using padstart here?
<https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/padStart>

File: src/BikramSambat.ts [hunk 6]
```diff
@@ -1,25 +1,83 @@
-// @ts-nocheck : TODO: Remove this line
-import {
-  NepaliDaysData,
-  NepaliMonthsData,
-  NewYearMappingData,
-  DaysInMonthsMappingData
-} from './data'
+import { format } from 'format'
+import { isDayValid, parse } from './parser'
```
Comment (aabishkaryal) [inline src/BikramSambat.ts]:
```suggestion
import { isDayValid, parse } from 'parser'
```

File: src/data/date-formats.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  | `${'YYYY'}${DateSeparator}${'MM'}`
+  | `${'YYYY' | 'YYY' | 'YY'}${DateSeparator}${'MM'}${DateSeparator}${'DD'}`
+  | `${'DD'}${DateSeparator}${'MMMM' | 'MM'}${DateSeparator}${'YYYY'}`
+  | `${'MMMM'}${DateSeparator}${'DD'}${DateSeparator}${'YYYY'}`
+  | `${'MM'}${DateSeparator}${'DD'}`
+
+interface DateFormat {
+  regex: RegExp
+  format: Format // General format of the date
+}
```
Comment (aabishkaryal) [inline src/data/date-formats.ts]:
```suggestion
export type DateFormat =
  | 'YYYY'
  | `${'YYYY'}${DateSeparator}${'MM'}`
  | `${'YYYY' | 'YYY' | 'YY'}${DateSeparator}${'MM'}${DateSeparator}${'DD'}`
  | `${'DD'}${DateSeparator}${'MMMM' | 'MM'}${DateSeparator}${'YYYY'}`
  | `${'MMMM'}${DateSeparator}${'DD'}${DateSeparator}${'YYYY'}`
  | `${'MM'}${DateSeparator}${'DD'}`

interface DateFormatDetails {
  regex: RegExp
  format: Format // General format of the date
}
```

or some better naming maybe?
we have format method and format type so could be confusing?

File: src/format.test.ts [hunk 8]
```diff
@@ -0,0 +1,175 @@
+import { format } from './format'
+import BikramSambat from './BikramSambat'
```
Comment (aabishkaryal) [inline src/format.test.ts]:
```suggestion
import { format } from 'format'
import BikramSambat from 'BikramSambat'
```

File: src/format.ts [hunk 9]
```diff
@@ -0,0 +1,48 @@
+import BikramSambat from 'BikramSambat'
+import { NepaliMonthsData, type Format } from 'data'
+import { generateDateFormatOrder } from 'utils/generateDateFormatOrder'
+
+const MonthNames = NepaliMonthsData.map((month) => month.en)
```
Comment (aabishkaryal) [inline src/format.ts]:
```suggestion
const MonthNamesEn = NepaliMonthsData.map((month) => month.en)
```

File: src/format.ts [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+/**
+ * Formats a BikramSambat date into the specified format.
+ *
+ * @param {BikramSambat} date - The BikramSambat date to be formatted.
+ * @param {Format} dateFormat - The desired format for formatting the date.
+ * @returns {string} - The formatted date string.
+ * @example <caption>Format a BikramSambat date into `YYYY-MM-DD` format.</caption>
+ * const date = new BikramSambat(2077, 1, 1)
```
Comment (aabishkaryal) [inline src/format.ts]:
```suggestion
 * const date = new BikramSambat(2077, 1, 1)
```
do we support this type of constructor? I don't think we do yet right?

File: src/parser.ts [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+ *
+ * @param {number} year - The year.
+ * @param {number} month - The month (1-12).
+ * @param {number} day - The day.
+ * @returns {boolean} `true` if the day is valid, otherwise `false`.
+ */
+export const isDayValid = (
+  year: number | undefined,
+  month: number | undefined,
+  day: number | undefined
```
Comment (aabishkaryal) [inline src/parser.ts]:
```suggestion
  year?: number,
  month?: number,
  day?: number
```

File: src/isDayValid.test.ts [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+  })
+
+  it('should return false for day out of range', () => {
+    expect(isDayValid(2075, 8, 35)).toBe(false)
+  })
+
+  it('should return false for day exceeding days in given month', () => {
+    expect(isDayValid(2075, 2, 32)).toBe(false)
+  })
+})
```
Comment (aabishkaryal) [inline src/isDayValid.test.ts:34]: can we add test case for:
```
isDayValid(undefined,undefined,undefined)
isDayValid(2070,undefined,undefined)
isDayValid(undefined,03,undefined)
isDayValid(undefined,undefined,23)
// and similar other combinations
```
since these are also valid inputs to the function.

File: src/parser.ts [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+      // checking if the date is valid
+      if (!isDayValid(year, month, day)) {
+        return 'Invalid Date'
+      }
+
+      return {
+        ...(year ? { year } : {}),
+        ...(month ? { month } : {}),
+        ...(day ? { day } : {})
+      }
```
Comment (aabishkaryal) [inline src/parser.ts]:
```suggestion
      const parsedDate: ParseResult = {}
      const order = generateDateFormatOrder(dateFormat.format)
      order.forEach((component, index) => {
        if (
          component === 'month' &&
          MonthNames.indexOf(match[index + 1]) >= 0
        ) {
          parsedDate[component] = MonthNames.indexOf(match[index + 1]) + 1
        } else {
          if (component === 'year' && match[index + 1].length < 4) {
            if (match[index + 1].length === 2)
              parsedDate[component] = parseInt('20' + match[index + 1])
            else parsedDate[component] = parseInt('2' + match[index + 1])
          } else {
            parsedDate[component] = parseInt(match[index + 1])
          }
        }
      })
      const { year, month, day } = parsedDate
      // checking if the date is valid
      if (!isDayValid(year, month, day)) {
        return 'Invalid Date'
      }

      return parsedDate
      ```

**Review by aabishkaryal — CHANGES_REQUESTED** [review]

File: src/parser.ts [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+ * @param {number} year - The year.
+ * @param {number} month - The month (1-12).
+ * @param {number} day - The day.
+ * @returns {boolean} `true` if the day is valid, otherwise `false`.
+ */
+export const isDayValid = (year?: number, month?: number, day?: number) => {
+  if (year === undefined || month === undefined || day === undefined) {
+    return false
+  }
+  if (month != undefined && (month < 1 || month > 12)) {
```
Comment (aabishkaryal) [inline src/parser.ts]:
```suggestion
  if (month< 1 || month > 12) {
```
since we know month is not going to be undefined at this point

File: src/parser.ts [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+    return false
+  }
+  if (year != undefined && (year > 2100 || year < 1975)) {
```
Comment (aabishkaryal) [inline src/parser.ts]:
```suggestion
  if  (year > 2100 || year < 1975) {
```

File: src/parser.ts [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+    return false
+  }
+
+  if (day != undefined && (day < 1 || day > 32)) {
```
Comment (aabishkaryal) [inline src/parser.ts]:
```suggestion
  if (day < 1 || day > 32) {
```

Comment (aabishkaryal) [reply src/parser.ts]: what about checking from the month in days data here? like some months never have 32 days even in those case those date are also invalid.

File: src/parser.ts [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+    return false
+  }
+
+  if (year != undefined && month != undefined && day != undefined) {
```
Comment (aabishkaryal) [inline src/parser.ts]: do we need this check anymore now?
since we return false if either is undefined

File: src/parser.ts [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+        parsedDate.day = 1
+      }
+
+      if (!isDayValid(parsedDate.year, parsedDate.month, parsedDate.day)) {
+        return 'Invalid Date'
+      }
+      return parsedDate
+    }
+  }
+  return 'Invalid Date'
```
Comment (aabishkaryal) [inline src/parser.ts]: what about having a constant for invalid date error since it is used in multiple places.

File: src/data/date-formats.ts [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+  regex: RegExp
+  format: DateFormat // General format of the date
+}
+
+/**
+ * Date formats supported by the parser and formatter
+ * @internal Used by the parser and formatter
+ * @ignore Exclude from docs
+ */
+export const DateFormats: DateFormatDetails[] = [
```
Comment (aabishkaryal) [inline src/data/date-formats.ts]:
```suggestion
export const DateFormatsDetails: DateFormatDetail[] = [
```

File: src/data/date-formats.ts [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+ * Date formats supported by the parser and formatter
+ */
+export type DateFormat =
+  | 'YYYY'
+  | `${'YYYY'}${DateSeparator}${'MM'}`
+  | `${'YYYY' | 'YYY' | 'YY'}${DateSeparator}${'MM'}${DateSeparator}${'DD'}`
+  | `${'DD'}${DateSeparator}${'MMMM' | 'MM'}${DateSeparator}${'YYYY'}`
+  | `${'MMMM'}${DateSeparator}${'DD'}${DateSeparator}${'YYYY'}`
+
+interface DateFormatDetails {
```
Comment (aabishkaryal) [inline src/data/date-formats.ts]:
```suggestion
interface DateFormatDetail {
```

File: src/format.test.ts [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+  it('should format date with mismatched separators (YYYY/MM DD)', () => {
+    const date = new BikramSambat('2079-8-15')
+    const result = format(date, 'YYYY/MM DD')
+    expect(result).toEqual('2079/08 15')
+  })
+  it('should format date with mismatched separators (MMMM DD YYYY)', () => {
+    const date = new BikramSambat('2079-8-15')
+    const result = format(date, 'MMMM DD YYYY')
+    expect(result).toEqual('Mangsir 15 2079')
+  })
```
Comment (aabishkaryal) [inline src/format.test.ts]: this doesn't have mismatched separators

File: src/parser.ts [hunk 22]
```diff
@@ -0,0 +1,90 @@
+import { NepaliMonthsData, DateFormats, DaysInMonthsMappingData } from './data'
+import { generateDateFormatOrder } from 'utils/generateDateFormatOrder'
+
+const MonthNames = NepaliMonthsData.map((month) => month.en)
```
Comment (aabishkaryal) [inline src/parser.ts]:
```suggestion
const MonthNamesEn = NepaliMonthsData.map((month) => month.en)
```

File: src/format.ts [hunk 23]
```diff
@@ -0,0 +1,48 @@
+import BikramSambat from 'BikramSambat'
+import { NepaliMonthsData, type DateFormat } from 'data'
+import { generateDateFormatOrder } from 'utils/generateDateFormatOrder'
+
+const MonthNamesEn = NepaliMonthsData.map((month) => month.en)
```
Comment (aabishkaryal) [inline src/format.ts]: maybe we could have it inside data since it is used in multiple places?

**Review by aabishkaryal — CHANGES_REQUESTED** [review]

**Review by aabishkaryal — APPROVED** [review]

**Review by ashiishme — APPROVED** [review]
I think there are still few things we might need to adjust but can be refactored later.
